### ISSUE TITLE
Add support for restriction value 'destination' for car

### ISSUE THREAD

**Issue by CoconutLimeMonkey** [graphhopper/graphhopper#733]
From: https://discuss.graphhopper.com/t/restrictedvalues-destination/475

The access restriction value destination is not regarded. A high penalty has to be applied when changing from an unrestricted way to a way with destination restriction.

Test case 1: Legal path avoids the hamlet Heitersberg and will make a ~20km detour.
https://graphhopper.com/maps/?point=47.414905%2C8.343494&point=47.415366%2C8.346804&locale=en-US&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale

Test case 2: Fastest path is through Bellikon 
https://graphhopper.com/maps/?point=47.36249%2C8.367419&point=47.415231%2C8.346503&locale=en-US&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale

**Comment by boldtrn** [comment]
Not sure if this belongs here, but I found a routing issue yesterday, that seems to happen, due to not correctly interpreted `destination`:
https://graphhopper.com/maps/?point=48.710638%2C9.450173&point=48.712171%2C9.453017&locale=de-DE&vehicle=car&weighting=fastest&elevation=true&use_miles=false&layer=Omniscale

This way should not be used:
http://www.openstreetmap.org/way/362442354#map=19/48.71141/9.45258&layers=ND

**Comment by karussell** [comment]
We currently ignore the destination thatswhy the issue. If we introduce this we would either need to introduce an additional priority (like we do e.g. for bikes) or make time estimates slightly incorrect.

**Comment by boldtrn** [comment]
Wouldn't be this priority exactly the right thing (speaking semantically)?
https://github.com/graphhopper/graphhopper/blob/ea1a8e207d02cb52f2b062712b2c7e9689830c5c/core/src/main/java/com/graphhopper/routing/util/PriorityCode.java#L30

**Comment by boldtrn** [comment]
I am currently looking into this issue.

IMHO, adding a penalty when switching from non-destination roads, to destination roads would not justify the the effort of implementation.

There are some other options I see right now:
- Implement it in the current CarFlagEncoder (probably the most helpful encoder for this right now)
 - We could modify speed to e.G. 5km/h (However, this would make the estimated travel times incorrect)
 - We could add a Priority to the CarFlagEncoder (probably not very nice, but the solution would be more correct)
- Implement it in the DataFlagEncoder, however this would require us to save the access tags of each highway and analyze them per vehicle (not supported yet)

Right now I guess, that the DataFlagEncoder solution would be the best in terms of future oriented development. The speed "trick" in the CarFlagEncoder would be a fast improvement for the current situation. Maybe we even want to implement both?

**Comment by karussell** [comment]
Yes, this sounds good

> We could add a Priority to the CarFlagEncoder (probably not very nice, but the solution would be more correct)

This is the correct solution, but we loose (like we do for bike&foot) that the weight is proportional to the time which is sometimes good for performance reasons, see #240 

So I vote to make your two proposed changes and potentially we need to revert the change in CarFlagEncoder e.g. if there are too many long ways with `destination`.

**Comment by michaz** [comment]
A strict implementation of this was just feature-requested on Facebook.

In a posting which I'm afraid I cannot link to at the moment because something is broken on Facebook. :-)

**Comment by karussell** [comment]
With 'strict' you mean that uses a priority difference and not via speed?

**Comment by boldtrn** [comment]
> A strict implementation of this was just feature-requested on Facebook.

The issue with this is, that we don't have enough context to create a wise decision here. Areas tagged with access=destination should be allowed to enter, but only if there is a waypoint in that area or the waypoint is only reachable using this road. AFAIK we cannot model this relation right now?

**Comment by michaz** [comment]
EDITed to 'stricter'. I think the strict meaning of the rule is "don't go there except if you destination is exactly on this road", and yes, I think we cannot model this right now.

**Comment by michaz** [comment]
Just wanted to pass on the interest in this feature.

**Comment by karussell** [comment]
> yes, I think we cannot model this right now

We could model it via priorities IMO (?) I.e. all destination tags get a very bad weighting that will only be used if necessary.

**Comment by michaz** [comment]
Infinite, even? Don't know exactly what a priority is (haven't looked at road routing much), but sounds good.

**Comment by karussell** [comment]
> Infinite, even?

This could make problems. E.g. in #729 we say that we want to consider the weighting to *exclude* edges based on infinite weights.

> Don't know exactly what a priority is (haven't looked at road routing much), but sounds good.

For bike routing we call the part of the flags this way and have different priorities like 1-7 to say it should be preferred without the need to change the speed

**Comment by michaz** [comment]
Ah, no, the reason is that with Infinity, all paths would be equally bad and the problem collapses. So almost infinity.

**Comment by boldtrn** [comment]
> Ah, no, the reason is that with Infinity, all paths would be equally bad and the problem collapses. So almost infinity.

You should be aware of the "Connection not found" issue, if the weight is exceeding a certain threshold, so I am not sure if this is practicable?

**Comment by michaz** [comment]
> You should be aware of the "Connection not found" issue, if the weight is exceeding a certain threshold, so I am not sure if this is practicable?

No, I'm not -- I'm not a road router at the moment. ,-) Tell me more, please.. What is that threshold and where does it come from?

**Comment by boldtrn** [comment]
@michaz, look for the `ConnectionNotFoundException`. I would assume this could happen for non road routing as well, but very unlikely, except there are different subnetworks. Also see the maxVisited nodes in the different `RoutingAlgorithm` classes. This could be one example where it might fail. Not sure if there are more cases @karussell?

**Comment by michaz** [comment]
> Ah I forgot, if we now set the weight very high, close to infinite, we most likely traverse the whole graph before even consider taking this road, this would make it a) very slow and b) might fail due to maxVisitedNode.

True, true, true.

**Comment by michaz** [comment]
And anything "properly multi-criteria" is super-slow and super-vulnerable to accidentally traversing the whole graph. Tricky.

**Comment by boldtrn** [comment]
> Ah I forgot, if we now set the weight very high, close to infinite, we most likely traverse the whole graph before even consider taking this road, this would make it a) very slow and b) might fail due to maxVisitedNode.

So if we want to do this strict, I think we have to get more context into the calculation, like checking every waypoint if it's only reachable from the outside without  using destination roads, if not, run a second shortest path search that allows destination roads, which could be still vulnerable for taking the wrong destination roads between two points, but if destination roads are also slow are have a bad priority it could work.

**Comment by michaz** [comment]
Or sneakily look at OSRM how they do it.

**Comment by karussell** [comment]
> Or sneakily look at OSRM how they do it

They do it via priorities, I think. Before they did it via manipulating the speed (ugly). But as they **only** use CH they won't get big performance issues I think.

**Comment by boldtrn** [comment]
@ftrebien, thanks for reporting, could you provide 1-2 example routes so we could have a look?

**Comment by ftrebien** [comment]
Sure. Here are some examples:
- [example 1](https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=-30.08305%2C-51.24681%3B-30.08196%2C-51.24401): through a mall
- [example 2](https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=-30.03992%2C-51.23297%3B-30.04263%2C-51.22884): through a government building
- [example 3](https://www.openstreetmap.org/directions?engine=graphhopper_bicycle&route=-30.00325%2C-51.12478%3B-30.00333%2C-51.12407): through a residential condominium

**Comment by boldtrn** [comment]
> Sure. Here are some examples

Thanks for these examples. These are great examples of routes that should not be taken. I will have a look and improve it for bike.

Maybe we should also improve this for foot then?

**Comment by boldtrn** [comment]
I created a PR for this #1272.

**Comment by ftrebien** [comment]
@boldtrn Sorry for the late reply. By moving the endpoints slightly north in examples 2 and 3, we get similar situations for pedestrian routing too. So I think this should be solved for foot as well.

**Comment by karussell** [comment]
PR #1752 does not fully fix this, especially for short edges. Maybe we use a weight offset, not (only) a factor?

**Comment by karussell** [comment]
We cannot just increase the factor because then situations like this:

https://graphhopper.com/maps/?point=51.528177%2C4.404659&point=51.523235%2C4.4135
![image](https://user-images.githubusercontent.com/129644/73576546-fb3d0300-447a-11ea-9295-39b824edbaa8.png)

will get worse. What happened here? The Boterstraat (left) and Boeiinksestraat (right) are both "destination"-only roads and distance is only slightly different. When we multiply the weight with 10:

```
weight=72 * 10 => 720
weight=63 * 10 => 630
```

an artificial difference is created (90), where the weight of Roosendaalsebaan (39) easily fits in, i.e. Roosendaalsebaan+Boeiinksestraat is then better than Boterstraat, although the way over Boterstraat is in all regards better (shorter & faster). The only "advantage" of the route over the Boeiinksestraat is that the time on a destination-only road is slightly smaller. 

**Update:** 

We could introduce some "barrier" edges (edges with high weight) before and after these roads. I.e. you put a penalty not directly on these edges but only if someone wants to go to them or leave them. The only downside could be if you start inside and end inside and for the shortest path you have to leave this "destination area" but this should be probably very rare or non-existent.

Implementation detail: instead of the barrier edges we could use turn costs and then such an avoidance can be created also per-request.

**Comment by karussell** [comment]
This is now implemented via turn costs.
